selfdrive/boardd/panda.cc

```diff
@@ -92,6 +87,64 @@ void Panda::cleanup() {
   }
 }
 
+int Panda::init_usb(libusb_context *context) {
+  int err = libusb_init(&context);
+  if (err != 0) {
+    LOGE("libusb initialization error");
+    return err;
+  }
+
+#if LIBUSB_API_VERSION >= 0x01000106
+  libusb_set_option(context, LIBUSB_OPTION_LOG_LEVEL, LIBUSB_LOG_LEVEL_INFO);
+#else
+  libusb_set_debug(context, 3);
+#endif
+
+  return err;
+}
+
+std::vector<std::string> Panda::list() {
+  // init libusb
+  ssize_t num_devices;
+  libusb_context *context = NULL;
+  libusb_device **dev_list = NULL;
+  std::vector<std::string> serials;
+  
+  int err = init_usb(context);
+  if (err != 0) { return serials; }
+
+  num_devices = libusb_get_device_list(context, &dev_list);
+  if (num_devices < 0) {
+    LOGE("libusb can't get device list");
+    goto finish;
+  }
+  for (size_t i = 0; i < num_devices; ++i) {
+    libusb_device *device = dev_list[i];
+    libusb_device_descriptor desc;
+    libusb_get_device_descriptor(device, &desc);
+    if (desc.idVendor == 0xbbaa && desc.idProduct == 0xddcc) {
+      libusb_device_handle *handle = NULL;
+      libusb_open(device, &handle);
+      unsigned char desc_serial[25];
+      int ret = libusb_get_string_descriptor_ascii(handle, desc.iSerialNumber, desc_serial, std::size(desc_serial));
+      libusb_release_interface(handle, 0);
+      libusb_close(handle);
+
+      if (ret < 0) { goto finish; }
+      serials.push_back(std::string(desc_serial, desc_serial+std::size(desc_serial)).c_str());
```

Comment:
```suggestion
      serials.push_back(std::string(desc_serial, ret));
```

Reply:
This is not gonna work.
error: no matching constructor for initialization of 'std::string' (aka 'basic_string<char>')
      serials.push_back(std::string(desc_serial, ret));

Reply:
I suppose because I am using unsigned char. Method I am using now is recommended in this case.

Reply:
Now you still rely on the null byte instead of the length returned by the libusb function. Just cast it to (char*) if needed.